### PR TITLE
`now dev` ssl support

### PR DESCRIPTION
**Pull request by krasimir**
## Why

Very often when developing locally you may need https instead of http. Issue: https://github.com/zeit/now-cli/issues/2380

## What

I've made two changes:
* `now dev` accepts `-c` and `-k` for the cert and key files
* run `https.createServer` instead of `http.createServer` if there are SSL options provided

I think it make sense to incorporate those changes into the `dev` command only since when you deploy you already have valid SSL working there.

## DAT

1. Create a key-cert pairs via 

```
openssl req -batch -new -newkey ec:<(openssl ecparam -name prime256v1) -nodes -keyout ca-key.pem -x509 -out ca.pem -days 3650 -subj "/CN=A localhost CA"
```

2. Build `now-cli` via `yarn build`
3. Jump to a folder that has `now.json` and place the generated key and cert files there.
4. Run `node <path to now-cli>/dist/index.js dev -c cert.pem -k key.pem`
5. Open https://localhost:3000 in your browser
6. You'll probably see a warning that the site is not secure. That's fine. Click "Advanced" and proceed forward.

### CODE REVIEW

**Comment by styfle** [comment]
Hi @krasimir Thanks for the PR!

I'm wondering what the use case is here.

The issue you linked to #2380 mentions use case for Slack which requires https on a public facing server.

This PR solves the https problem but it is still running locally (not publicly) so you would have to use something like ngrok.

We have discussed using a flag like `now dev --cloud` for this particular use case but I would defer to @leo and @TooTallNate for how to solve this problem.

File: src/commands/dev/dev.ts [hunk 1]
```diff
@@ -26,7 +29,9 @@ export default async function dev(
   const cwd = path.join(process.cwd(), dir);
   const port = opts['-p'] || opts['--port'];
   const debug = opts['-d'] || opts['--debug'];
-  const devServer = new DevServer(cwd, { output, debug });
+  const cert = opts['-c'] || opts['--cert'];
+  const key = opts['-k'] || opts['--key'];
```
Comment (styfle) [inline src/commands/dev/dev.ts:34]: Let's avoid using short hand since we may wish to utilize `-c` or `-k` for something else in the future.

**Comment by krasimir** [comment]
Hey @styfle, I'm a little bit confused by 
> The issue you linked to #2380 mentions use case for Slack which requires https on a public facing server.

@alexluong mentions `to set up https server with command now dev` so I was thinking that he/she is talking about running his/her project locally and not able to communicate with Slack. 

Anyways, it is very possible that I misunderstood the issue :) So feel free to close this PR.

File: src/commands/dev/dev.ts (hunk 1 above)
Comment (krasimir) [reply src/commands/dev/dev.ts:34]: I agree. Other options for passing this information:
* Putting this paths as configuration in `now.json`
* Env variables
* Maybe more specific --ssl and we ditch `-c` and `-k` that I added above

I have the feeling that we have to close this PR but if not I'll appreciate your advice on this. I like the minimalistic API that `now` has and would love to keep it as it is. Just wondering what's the most feasible approach.

**Comment by styfle** [comment]
@krasimir If you take a look at https://ngrok.com you'll see the headline says "Public URLs for exposing your local web server."

I believe this is what the user wants since Slack Hooks need to call a URL and it can't call your local dev machine without a public IP address, forwarded ports, etc.

This is a something that `ngrok` solves for the user but it would be nice to have this built into Now.

**Comment by ywg-jean** [comment]
I believe it is still useful to have https for locally run dev instances. : 
- local testing with HTTP hides any mixed content issues that can break a production HTTPS website
- some browser features are only available in [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts) 

in my case it would also eliminate the http warnings from stripe.js

**Comment by styfle** [comment]
@ywg-jean You convinced me. I forgot about Service Workers 😅 

@TooTallNate @leo Are you good with this PR?

**Comment by ywg-jean** [comment]
As a bonus if you need valid ssl certs for localhost, 127.0.0.1 or other custom dev domains, [mkcert](https://github.com/FiloSottile/mkcert) is a really nice tool.

**Comment by amwmedia** [comment]
this would be a big help for me too 👍

**Comment by ctrlplusb** [comment]
Came across this post detailing how to easily create localhost certs.

https://dev.to/rhymes/really-easy-way-to-use-https-on-localhost-341m

Some may find it useful

**Comment by denkristoffer** [comment]
Another use case: Microsoft Office add-ins have to be served over HTTPS. This PR could really speed up development by removing the need to deploy every time 👍

**Comment by leo** [comment]
Question for everyone in here: Would it help to instead have a flag like `--live`, which creates a HTTPS URL to which you can send requests. Those requests will then be piped to your local device and to into `now dev`.

Does that cover your use case? What are the cases in which you need SSL for `now dev`?

**Comment by denkristoffer** [comment]
@leo Yes I think a flag like `--live` would cover my use case perfectly.

**Comment by iamstarkov** [comment]
@Leo when you say HTTPS URL do you mean it will be public deployment tunneled to local env?

**Comment by leo** [comment]
> @leo when you say HTTPS URL do you mean it will be public deployment tunneled to local env?

Yes, @iamstarkov.

**Comment by iamstarkov** [comment]
@leo will it cover use case when you need https for [features available only in secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts/features_restricted_to_secure_contexts)? e.g. push, service workers, bluetooth, sensors, etc

On the other note, does the tunnel expose local dev environment in insecure way?

**Comment by leo** [comment]
@iamstarkov We'll make sure this won't happen! 👍

**Comment by iamstarkov** [comment]
@Leo is `--live` out?

**Comment by ibraelillo** [comment]
hey?? this pr wont pass??

**Comment by leo** [comment]
We just cleaned up the repository by deleting all our branches and re-creating only one of them. This now allows us to create canary and stable releases without having to deal with two separate branches.

As a result, this PR was automatically closed (not manually, even though the UI indicates that).

The team is very grateful for your contribution! However, we've decided not to re-open it as we'll be looking into implementing the `--live` flag very soon instead.

Thank you so much! 🙏

**Comment by liamdanielduffy** [comment]
@leo 

I noticed that this PR was closed because the `--live` flag was expected to be implemented soon at the time of the last comment.

However, [this comment](https://github.com/vercel/vercel/issues/3798) suggests that the `--live` flag will not be implemented. That seems to make this PR relevant again.

I also want to note that while [the comment linked above](https://github.com/vercel/vercel/issues/3798) suggests pushing to a live preview deployment is an adequate solution for testing a site over HTTPS, this is a much slower feedback loop than seeing changes live in localhost a second or two after making them.

Vercel does such an awesome job on DX. This would make locally developing features that require HTTPS or a live URL to work way easier!

(My use case is building a checkout with Apple Pay, via Stripe's [Payment Request Button](https://stripe.com/docs/stripe-js/elements/payment-request-button))

**Pull request closed without merging**